Initial draft for Java quickstart (#6142)
github/docs: content/codespaces/getting-started-with-codespaces/getting-started-with-your-java-project-in-codespaces.md

---
title: Getting started with your Java project in Codespaces
shortTitle: Getting started with your Java project
intro: 'You can create a custom dev container with all the tools necessary to get started with your Java project in {% data variables.product.prodname_codespaces %}.'
product: '{% data reusables.gated-features.codespaces %}'
versions:
  free-pro-team: '*'
topics:
  - Codespaces
---

### Introduction

> This guide shows you how to set up your Java project in {% data variables.product.prodname_codespaces %}. It will take you through an example of opening your project in a codespace,and adding and modifying a dev container configuration from a template.

Reviewer: ```suggestion
This guide shows you how to set up your Java project in {% data variables.product.prodname_codespaces %}. It will take you through an example of opening your project in a codespace, and adding and modifying a dev container configuration from a template.
```

#### Prerequisites 

- You should have an existing Java project in a repository on {% data variables.product.prodname_dotcom_the_website %}. If you don't have a project, you can try this tutorial with the following example: https://github.com/microsoft/vscode-remote-try-java
- You must have {% data variables.product.prodname_codespaces %} enabled for your organization.

### Step 1: Open your project in a codespace

1. Navigate to your project's repository. Use the {% octicon "download" aria-label="The download icon" %} **Code** drop-down menu, and select **Open with Codespaces**. If you don’t see this option, your project isn’t available for {% data variables.product.prodname_codespaces %}.
  
  ![Open with Codespaces button](/assets/images/help/codespaces/open-with-codespaces-button.png) 

2. To create a new codespace, click {% octicon "plus" aria-label="The plus icon" %} **New codespace**.
  ![New codespace button](/assets/images/help/codespaces/new-codespace-button.png)

> When you create a codespace, your project is created on a remote VM that is dedicated to you. By default, the container for your codespace has many languages and runtimes including Node.js, JavaScript, Typescript, Java, nvm, npm, and yarn. It also includes a common set of tools like git, wget, rsync, openssh, and nano. 

Reviewer: ```suggestion
When you create a codespace, your project is created on a remote VM that is dedicated to you. By default, the container for your codespace has many languages and runtimes including Java, nvm, npm, and yarn. It also includes a common set of tools like git, wget, rsync, openssh, and nano. 
```

// https://github.com/microsoft/vscode-dev-containers/tree/v0.159.0/containers/java
{
	"name": "Java",
	"build": {
		"dockerfile": "Dockerfile",
		"args": {
			// Update the VARIANT arg to pick a Java version: 11, 14
			"VARIANT": "11",
			// Options
			"INSTALL_MAVEN": "true",
			"INSTALL_GRADLE": "false",
			"INSTALL_NODE": "false",
			"NODE_VERSION": "lts/*"
		}
	},

	// Set *default* container specific settings.json values on container create.
	"settings": { 
		"terminal.integrated.shell.linux": "/bin/bash",
		"java.home": "/docker-java-home",
		"maven.executable.path": "/usr/local/sdkman/candidates/maven/current/bin/mvn"
	},
	
	// Add the IDs of extensions you want installed when the container is created.
	"extensions": [
		"vscjava.vscode-java-pack"
	],

	// Use 'forwardPorts' to make a list of ports inside the container available locally.
	// "forwardPorts": [],

	// Use 'postCreateCommand' to run commands after the container is created.
	// "postCreateCommand": "java -version",

	// Uncomment to connect as a non-root user. See https://aka.ms/vscode-remote/containers/non-root.
	"remoteUser": "vscode"
}
```

> - **Name** - We can name our dev container anything, this is just the default
> - **Build** - Our build properties
>   - **Dockerfile** - In our build object, Dockerfile is a reference to the Dockerfile in the same folder that was the second file added to our project. This is the reference path.
>   - **Args**
>     - **Variant**: We only have one build argument here which is the node variant we want to use which is passed into our Dockerfile.
> - **Settings** - These are {% data variables.product.prodname_vscode %} settings.
>   - **Terminal.integrated.shell.linux** - While bash is the default here, you could use other terminal shells by modifying this.
> - **Extensions** - These are extensions included by default.
>   - **Vscjava.vscode-java-pack** - The Java Extension Pack provides popular extensions for Java development to get you started.
> - **forwardPorts** - By default we can forward a port, like port 3000, but these will also forward automatically.
> - **postCreateCommand** - If we want to run anything after we land in our codespace that’s not defined in our Dockerfile, like yarn install or npm install, we can do that here.
> - **remoteUser** - We’re running as the `vscode` user, but you can optionally set this to `root`.

Reviewer: ```suggestion
- **Name** - You can name your dev container anything, this is just the default.
- **Build** - The build properties.
  - **Dockerfile** - In the build object, dockerfile is a reference to the Dockerfile that was also added from the template.
  - **Args**
    - **Variant**: This file only contains one build argument, which is the Java version that is passed into the Dockerfile.
- **Settings** - These are {% data variables.product.prodname_vscode %} settings that you can set.
  - **Terminal.integrated.shell.linux** - While bash is the default here, you could use other terminal shells by modifying this.
- **Extensions** - These are extensions included by default.
  - **Vscjava.vscode-java-pack** - The Java Extension Pack provides popular extensions for Java development to get you started.
- **forwardPorts** - Any ports listed here will be forwarded automatically.
- **postCreateCommand** - If you want to run anything after you land in your codespace that’s not defined in the Dockerfile, you can do that here.
- **remoteUser** - By default, you’re running as the `vscode` user, but you can optionally set this to `root`.
```

3. To rebuild your container, access the command palette (`shift command P` / `shift control P`), then start typing "rebuild". Click **Codespaces: Rebuild Container**. 

  ![Rebuild container option](/assets/images/help/codespaces/codespaces-rebuild.png)

  Rebuilding inside your codespace ensures your changes work as expected before you commit the changes to the repository. If something does result in a failure, you’ll be placed in a codespace with a recovery container that you can rebuild from to keep adjusting your container.

### Step 4: Run your application

In the previous section, you used the `postCreateCommand` to install a set of packages via npm. You can now use this to run our application with npm.

1. Run your application by pressing `F5`.

2. When your project starts, you should see a toast in the bottom right corner with a prompt to connect to the port your project uses. 

  ![Port forwarding toast](/assets/images/help/codespaces/codespaces-port-toast.png)

### Step 5: Commit your changes

> Once you've made changes to your codespace, either new code or configuration changes, you'll want to commit your changes. Committing changes to your repository ensures that anyone else who creates a codespace from this repository has the same configuration. This also means that any customization you do, such as adding {% data variables.product.prodname_vscode %} extensions, will appear for all users.
> 
> 1. In the Activity Bar, click on the **Source Control** view.
> 
>   ![Source control view](/assets/images/help/codespaces/codespaces-commit-activity.png)  
> 
> 2. To stage your changes, click  **+**.
> 
>   ![Explorer with staging button highlighted](/assets/images/help/codespaces/codespaces-commit-stage.png)
> 
> 3. Type a commit message and then use `Ctrl+Enter` / `cmd+Enter` to commit the changes.  
> 
>   ![Explorer with commit message added](/assets/images/help/codespaces/codespaces-commit-commit-message.png)  
> 
> 4. To create a PR, click the PR icon. 
> 
>   ![Explorer with staging button highlighted](/assets/images/help/codespaces/codespaces-commit-pr-button.png)  
> 
> 5. Select the branches that you want to merge into, then click **Create**.
> 
>   ![Explorer with staging button highlighted](/assets/images/help/codespaces/codespaces-commit-pr.png)  

Reviewer: ```suggestion
{% data reusables.codespaces.committing-link-to-procedure %}
```
